NPE in :quarkusTestConfig gradle task with 1.3.0.Alpha2

**Describe the bug**
Running :test task explodes with NPE after update to 1.3.0.Alpha2


**Expected behavior**
Would expect test execution to work as it does with 1.2.1.Final

**Actual behavior**
NPE, see stack trace at end of report.

**To Reproduce**
Steps to reproduce the behavior:
1. Clone https://github.com/jskov/dns-filter
2. Update quarkus gradle plugin (in build.gradle) and quarkus (in gradle.properties) to 1.3.0.Alpha2
3. ./gradlew test

**Configuration**
Gradle 6.2, running on Fedora 31 with java 11 or 13

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`:  Linux thinktwice.mada.dk 5.4.18-200.fc31.x86_64 #1 SMP Fri Feb 7 14:50:05 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
 - Output of `java -version`: OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.5+10, mixed mode)
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 1.3.0.Alpha2

**Additional context**
Full stack trace:

> * Exception is:
> org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':quarkusTestConfig'.
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.lambda$executeIfValid$1(ExecuteActionsTaskExecuter.java:205)
> 	at org.gradle.internal.Try$Failure.ifSuccessfulOrElse(Try.java:263)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:203)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:184)
> 	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
> 	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
> 	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
> 	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
> 	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
> 	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
> 	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
> 	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
> 	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:416)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:406)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:102)
> 	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
> 	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
> 	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
> 	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:372)
> 	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:359)
> 	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:352)
> 	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:338)
> 	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
> 	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
> 	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
> 	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
> 	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
> 	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
> 	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
> Caused by: java.lang.IllegalStateException: Failed to resolve deployment classpath
> 	at io.quarkus.gradle.tasks.QuarkusTestConfig.setupTest(QuarkusTestConfig.java:38)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:104)
> 	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:49)
> 	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:42)
> 	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:28)
> 	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:727)
> 	at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:694)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$3.run(ExecuteActionsTaskExecuter.java:568)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:402)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:394)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:165)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:250)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:158)
> 	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:92)
> 	at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:553)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:536)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.access$300(ExecuteActionsTaskExecuter.java:109)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.executeWithPreviousOutputFiles(ExecuteActionsTaskExecuter.java:276)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$TaskExecution.execute(ExecuteActionsTaskExecuter.java:265)
> 	at org.gradle.internal.execution.steps.ExecuteStep.lambda$execute$1(ExecuteStep.java:33)
> 	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:33)
> 	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:26)
> 	at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:67)
> 	at org.gradle.internal.execution.steps.CleanupOutputsStep.execute(CleanupOutputsStep.java:36)
> 	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:49)
> 	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:34)
> 	at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:43)
> 	at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:73)
> 	at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:54)
> 	at org.gradle.internal.execution.steps.CatchExceptionStep.execute(CatchExceptionStep.java:34)
> 	at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:44)
> 	at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:54)
> 	at org.gradle.internal.execution.steps.SnapshotOutputsStep.execute(SnapshotOutputsStep.java:38)
> 	at org.gradle.internal.execution.steps.BroadcastChangingOutputsStep.execute(BroadcastChangingOutputsStep.java:49)
> 	at org.gradle.internal.execution.steps.CacheStep.executeWithoutCache(CacheStep.java:153)
> 	at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:67)
> 	at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:41)
> 	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:44)
> 	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:33)
> 	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
> 	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
> 	at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:92)
> 	at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:85)
> 	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
> 	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
> 	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
> 	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
> 	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
> 	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
> 	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:94)
> 	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
> 	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:79)
> 	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:53)
> 	at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
> 	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:78)
> 	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:78)
> 	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
> 	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:39)
> 	at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:40)
> 	at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:28)
> 	at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
> 	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:192)
> 	... 30 more
> Caused by: java.lang.NullPointerException
> 	at io.quarkus.bootstrap.model.AppArtifact.writeObject(AppArtifact.java:40)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at io.quarkus.gradle.tasks.QuarkusGradleUtils.serializeAppModel(QuarkusGradleUtils.java:15)
> 	at io.quarkus.gradle.tasks.QuarkusTestConfig.setupTest(QuarkusTestConfig.java:30)
> 	... 95 more
>

With #7353 it fails with
```
SimpleDnsQueryTest > testDnsLookup() FAILED
    java.lang.RuntimeException at QuarkusTestExtension.java:164
        Caused by: java.lang.NoClassDefFoundError at Class.java:-2
            Caused by: java.lang.ClassNotFoundException at BuiltinClassLoader.java:581
```
It works with Quarkus 1.2.1.Final though.

I don't think #7353 causes that tbh, but i may be wrong :)

Em Sáb, 22 de fev de 2020 18:33, Alexey Loubyansky <notifications@github.com>
escreveu:

> With #7353 <https://github.com/quarkusio/quarkus/pull/7353> it fails with
>
> SimpleDnsQueryTest > testDnsLookup() FAILED
>     java.lang.RuntimeException at QuarkusTestExtension.java:164
>         Caused by: java.lang.NoClassDefFoundError at Class.java:-2
>             Caused by: java.lang.ClassNotFoundException at BuiltinClassLoader.java:581
>
> It works with Quarkus 1.2.1.Final though.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/quarkusio/quarkus/issues/7357?email_source=notifications&email_token=AAANG5M4A7WKQPTAA4FEEATREGK2LA5CNFSM4KZP7ZQ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEMVLCHQ#issuecomment-590000414>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAANG5NIBWH5DTG5UHDYMILREGK2LANCNFSM4KZP7ZQQ>
> .
>

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/AppModelGradleResolver.java
@@ -7,9 +7,11 @@ import java.nio.file.FileSystem;
 import java.nio.file.FileSystems;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
@@ -19,6 +21,7 @@ import org.gradle.api.GradleException;
 import org.gradle.api.Project;
 import org.gradle.api.artifacts.Configuration;
 import org.gradle.api.artifacts.Dependency;
+import org.gradle.api.artifacts.DependencySet;
 import org.gradle.api.artifacts.ModuleIdentifier;
 import org.gradle.api.artifacts.ModuleVersionIdentifier;
 import org.gradle.api.artifacts.ResolvedArtifact;
@@ -27,8 +30,11 @@ import org.gradle.api.file.RegularFile;
 import org.gradle.api.internal.artifacts.DefaultModuleIdentifier;
 import org.gradle.api.internal.artifacts.dependencies.DefaultDependencyArtifact;
 import org.gradle.api.internal.artifacts.dependencies.DefaultExternalModuleDependency;
+import org.gradle.api.plugins.Convention;
 import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.plugins.JavaPluginConvention;
 import org.gradle.api.provider.Provider;
+import org.gradle.api.tasks.SourceSet;
 import org.gradle.jvm.tasks.Jar;
 
 import io.quarkus.bootstrap.BootstrapConstants;
@@ -38,6 +44,7 @@ import io.quarkus.bootstrap.model.AppDependency;
 import io.quarkus.bootstrap.model.AppModel;
 import io.quarkus.bootstrap.resolver.AppModelResolver;
 import io.quarkus.bootstrap.resolver.AppModelResolverException;
+import io.quarkus.gradle.tasks.QuarkusGradleUtils;
 import io.quarkus.runtime.LaunchMode;
 
 public class AppModelGradleResolver implements AppModelResolver {
@@ -124,23 +131,24 @@ public class AppModelGradleResolver implements AppModelResolver {
         if (appModel != null && appModel.getAppArtifact().equals(appArtifact)) {
             return appModel;
         }
-        final List<Dependency> extensionDeps = new ArrayList<>();
+        final List<Dependency> directExtensionDeps = new ArrayList<>();
         final List<AppDependency> userDeps = new ArrayList<>();
         Map<AppArtifactKey, AppDependency> versionMap = new HashMap<>();
         Map<ModuleIdentifier, ModuleVersionIdentifier> userModules = new HashMap<>();
 
-        collectDependencies(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME, appBuilder, extensionDeps, userDeps, versionMap,
+        collectDependencies(JavaPlugin.COMPILE_CLASSPATH_CONFIGURATION_NAME, appBuilder, directExtensionDeps, userDeps,
+                versionMap,
                 userModules);
         if (mode == LaunchMode.TEST) {
-            collectDependencies(JavaPlugin.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME, appBuilder, extensionDeps, userDeps,
+            collectDependencies(JavaPlugin.TEST_COMPILE_CLASSPATH_CONFIGURATION_NAME, appBuilder, directExtensionDeps, userDeps,
                     versionMap, userModules);
         }
 
         final List<AppDependency> deploymentDeps = new ArrayList<>();
         final List<AppDependency> fullDeploymentDeps = new ArrayList<>();
-        if (!extensionDeps.isEmpty()) {
+        if (!directExtensionDeps.isEmpty()) {
             final Configuration deploymentConfig = project.getConfigurations()
-                    .detachedConfiguration(extensionDeps.toArray(new Dependency[0]));
+                    .detachedConfiguration(directExtensionDeps.toArray(new Dependency[0]));
             final ResolvedConfiguration rc = deploymentConfig.getResolvedConfiguration();
             for (ResolvedArtifact a : rc.getResolvedArtifacts()) {
                 final ModuleVersionIdentifier userVersion = userModules.get(getModuleId(a));
@@ -150,26 +158,41 @@ public class AppModelGradleResolver implements AppModelResolver {
                 final AppDependency dependency = toAppDependency(a);
                 fullDeploymentDeps.add(dependency);
                 if (!userDeps.contains(dependency)) {
-                    deploymentDeps.add(alignVersion(dependency, versionMap));
+                    AppDependency deploymentDep = alignVersion(dependency, versionMap);
+                    deploymentDeps.add(deploymentDep);
                 }
             }
         }
         fullDeploymentDeps.addAll(userDeps);
 
-        // In the case of quarkusBuild (which is the primary user of this),
-        // it's not necessary to actually resolve the original application JAR
         if (!appArtifact.isResolved()) {
             final Jar jarTask = (Jar) project.getTasks().findByName(JavaPlugin.JAR_TASK_NAME);
             if (jarTask == null) {
                 throw new AppModelResolverException("Failed to locate task 'jar' in the project.");
             }
             final Provider<RegularFile> jarProvider = jarTask.getArchiveFile();
+            Path classesDir = null;
             if (jarProvider.isPresent()) {
                 final File f = jarProvider.get().getAsFile();
                 if (f.exists()) {
-                    appArtifact.setPath(f.toPath());
+                    classesDir = f.toPath();
                 }
             }
+            if (classesDir == null) {
+                final Convention convention = project.getConvention();
+                JavaPluginConvention javaConvention = convention.findPlugin(JavaPluginConvention.class);
+                if (javaConvention != null) {
+                    final SourceSet mainSourceSet = javaConvention.getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                    final String classesPath = QuarkusGradleUtils.getClassesDir(mainSourceSet, jarTask);
+                    if (classesPath != null) {
+                        classesDir = Paths.get(classesPath);
+                    }
+                }
+            }
+            if (classesDir == null) {
+                throw new AppModelResolverException("Failed to locate classes directory for " + appArtifact);
+            }
+            appArtifact.setPath(classesDir);
         }
         appBuilder.addRuntimeDeps(userDeps)
                 .addFullDeploymentDeps(fullDeploymentDeps)
@@ -178,10 +201,14 @@ public class AppModelGradleResolver implements AppModelResolver {
         return this.appModel = appBuilder.build();
     }
 
-    private void collectDependencies(String configName, AppModel.Builder appBuilder, final List<Dependency> extensionDeps,
+    private void collectDependencies(String configName, AppModel.Builder appBuilder, final List<Dependency> directExtensionDeps,
             final List<AppDependency> userDeps, Map<AppArtifactKey, AppDependency> versionMap,
             Map<ModuleIdentifier, ModuleVersionIdentifier> userModules) {
         final Configuration config = project.getConfigurations().getByName(configName);
+
+        final DependencySet incomingSet = config.getIncoming().getDependencies();
+        final Set<AppArtifactKey> directDeps = new HashSet<>(incomingSet.size());
+        incomingSet.forEach(d -> directDeps.add(new AppArtifactKey(d.getGroup(), d.getName())));
         for (ResolvedArtifact a : config.getResolvedConfiguration().getResolvedArtifacts()) {
             if (!isDependency(a)) {
                 continue;
@@ -204,8 +231,10 @@ public class AppModelGradleResolver implements AppModelResolver {
                     throw new GradleException("Failed to process " + f, e);
                 }
             }
-            if (dep != null) {
-                extensionDeps.add(dep);
+            if (dep != null &&
+                    directDeps.contains(new AppArtifactKey(dependency.getArtifact().getGroupId(),
+                            dependency.getArtifact().getArtifactId()))) {
+                directExtensionDeps.add(dep);
             }
         }
     }

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPluginExtension.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/QuarkusPluginExtension.java
@@ -1,15 +1,23 @@
 package io.quarkus.gradle;
 
 import java.io.File;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Set;
 
 import org.gradle.api.Project;
 import org.gradle.api.file.FileCollection;
+import org.gradle.api.file.RegularFile;
+import org.gradle.api.plugins.Convention;
+import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.plugins.JavaPluginConvention;
+import org.gradle.api.provider.Provider;
 import org.gradle.api.tasks.SourceSet;
+import org.gradle.jvm.tasks.Jar;
 
 import io.quarkus.bootstrap.model.AppArtifact;
 import io.quarkus.bootstrap.resolver.AppModelResolver;
+import io.quarkus.gradle.tasks.QuarkusGradleUtils;
 import io.quarkus.runtime.LaunchMode;
 
 public class QuarkusPluginExtension {
@@ -28,6 +36,36 @@ public class QuarkusPluginExtension {
 
     public QuarkusPluginExtension(Project project) {
         this.project = project;
+    }
+
+    public Path appJarOrClasses() {
+        final Jar jarTask = (Jar) project.getTasks().findByName(JavaPlugin.JAR_TASK_NAME);
+        if (jarTask == null) {
+            throw new RuntimeException("Failed to locate task 'jar' in the project.");
+        }
+        final Provider<RegularFile> jarProvider = jarTask.getArchiveFile();
+        Path classesDir = null;
+        if (jarProvider.isPresent()) {
+            final File f = jarProvider.get().getAsFile();
+            if (f.exists()) {
+                classesDir = f.toPath();
+            }
+        }
+        if (classesDir == null) {
+            final Convention convention = project.getConvention();
+            JavaPluginConvention javaConvention = convention.findPlugin(JavaPluginConvention.class);
+            if (javaConvention != null) {
+                final SourceSet mainSourceSet = javaConvention.getSourceSets().getByName(SourceSet.MAIN_SOURCE_SET_NAME);
+                final String classesPath = QuarkusGradleUtils.getClassesDir(mainSourceSet, jarTask);
+                if (classesPath != null) {
+                    classesDir = Paths.get(classesPath);
+                }
+            }
+        }
+        if (classesDir == null) {
+            throw new RuntimeException("Failed to locate project's classes directory");
+        }
+        return classesDir;
     }
 
     public File outputDirectory() {

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusBuild.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusBuild.java
@@ -15,7 +15,6 @@ import io.quarkus.bootstrap.app.CuratedApplication;
 import io.quarkus.bootstrap.app.QuarkusBootstrap;
 import io.quarkus.bootstrap.model.AppArtifact;
 import io.quarkus.bootstrap.resolver.AppModelResolver;
-import io.quarkus.bootstrap.resolver.AppModelResolverException;
 
 public class QuarkusBuild extends QuarkusTask {
 
@@ -54,13 +53,9 @@ public class QuarkusBuild extends QuarkusTask {
         getLogger().lifecycle("building quarkus runner");
 
         final AppArtifact appArtifact = extension().getAppArtifact();
+        appArtifact.setPath(extension().appJarOrClasses());
         final AppModelResolver modelResolver = extension().getAppModelResolver();
-        try {
-            // this needs to be done otherwise the app artifact doesn't get a proper path
-            modelResolver.resolveModel(appArtifact);
-        } catch (AppModelResolverException e) {
-            throw new GradleException("Failed to resolve application model " + appArtifact + " dependencies", e);
-        }
+
         final Properties realProperties = getBuildSystemProperties(appArtifact);
 
         boolean clear = false;

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java
@@ -17,7 +17,6 @@ import java.net.Socket;
 import java.net.URI;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
-import java.nio.file.Files;
 import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Arrays;
@@ -30,7 +29,6 @@ import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
-import java.util.stream.Stream;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipOutputStream;
 
@@ -42,7 +40,6 @@ import org.gradle.api.artifacts.Dependency;
 import org.gradle.api.artifacts.DependencySet;
 import org.gradle.api.artifacts.ProjectDependency;
 import org.gradle.api.artifacts.ResolvedDependency;
-import org.gradle.api.file.FileCollection;
 import org.gradle.api.plugins.Convention;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.plugins.JavaPluginConvention;
@@ -293,51 +290,13 @@ public class QuarkusDev extends QuarkusTask {
                     sourcePaths.add(sourceDir.getAbsolutePath());
                 }
 
-                FileCollection classesDirs = mainSourceSet.getOutput().getClassesDirs();
-                final String classesPaths;
-                Set<File> classDirFiles = classesDirs.getFiles();
-                if (classDirFiles.size() == 1) {
-                    classesPaths = classesDirs.getAsPath();
-                } else {
-                    //there does not seem to be any sane way of dealing with multiple output dirs, as there does not seem
-                    //to be a way to map them. We will need to address this at some point, but for now we just stick them
-                    //all in a temp dir
-
-                    Path path = getTemporaryDir().toPath();
-                    classesPaths = path.toAbsolutePath().toString();
-                    for (File c : classDirFiles) {
-                        Path cd = c.toPath();
-                        if (!Files.exists(cd)) {
-                            continue;
-                        }
-                        try (Stream<Path> stream = Files.walk(cd)) {
-                            stream.forEach(s -> {
-                                try {
-                                    if (Files.isDirectory(s)) {
-                                        return;
-                                    }
-                                    final Path file = cd.relativize(path);
-                                    final Path targetPath = path.resolve(file.toString());
-                                    Files.createDirectories(targetPath.getParent());
-                                    byte[] data = Files.readAllBytes(s);
-                                    Files.write(targetPath, data);
-                                } catch (IOException e) {
-                                    throw new RuntimeException(e);
-                                }
-
-                            });
-                        }
-                    }
-
-                }
-
                 String resourcePaths = mainSourceSet.getResources().getSourceDirectories().getSingleFile().getAbsolutePath(); //TODO: multiple resource directories
 
                 DevModeContext.ModuleInfo wsModuleInfo = new DevModeContext.ModuleInfo(
                         dependencyProject.getName(),
                         dependencyProject.getProjectDir().getAbsolutePath(),
                         sourcePaths,
-                        classesPaths,
+                        QuarkusGradleUtils.getClassesDir(mainSourceSet, this),
                         resourcePaths);
 
                 context.getModules().add(wsModuleInfo);
@@ -366,7 +325,10 @@ public class QuarkusDev extends QuarkusTask {
             final String outputResourcesDirectory = extension.outputConfigDirectory().getAbsolutePath();
             context.getClassesRoots().add(extension.outputDirectory().getAbsoluteFile());
             if (!outputClassDirectory.equals(outputResourcesDirectory)) {
-                context.getClassesRoots().add(extension.outputConfigDirectory().getAbsoluteFile());
+                final File configDir = extension.outputConfigDirectory().getAbsoluteFile();
+                if (configDir.exists()) {
+                    context.getClassesRoots().add(configDir);
+                }
             }
             context.setFrameworkClassesDir(wiringClassesDirectory.getAbsoluteFile());
             context.setCacheDir(new File(getBuildDir(), "transformer-cache").getAbsoluteFile());
@@ -401,7 +363,7 @@ public class QuarkusDev extends QuarkusTask {
                 out.write(bytes.toByteArray());
             }
 
-            final Path serializedModel = QuarkusGradleUtils.serializeAppModel(appModel);
+            final Path serializedModel = QuarkusGradleUtils.serializeAppModel(appModel, this);
             serializedModel.toFile().deleteOnExit();
             args.add("-D" + BootstrapConstants.SERIALIZED_APP_MODEL + "=" + serializedModel.toAbsolutePath());
 

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusGradleUtils.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusGradleUtils.java
@@ -1,19 +1,80 @@
 package io.quarkus.gradle.tasks;
 
+import java.io.File;
 import java.io.IOException;
 import java.io.ObjectOutputStream;
+import java.io.UncheckedIOException;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+
+import org.gradle.api.file.FileCollection;
+import org.gradle.api.internal.AbstractTask;
+import org.gradle.api.tasks.SourceSet;
 
 import io.quarkus.bootstrap.model.AppModel;
+import io.quarkus.bootstrap.util.IoUtils;
 
 public class QuarkusGradleUtils {
 
-    public static Path serializeAppModel(final AppModel appModel) throws IOException {
-        final Path serializedModel = Files.createTempFile("quarkus-", "-gradle-test");
+    private static final String ERROR_COLLECTING_PROJECT_CLASSES = "Failed to collect project's classes in a temporary dir";
+
+    public static Path serializeAppModel(final AppModel appModel, AbstractTask context) throws IOException {
+        final Path serializedModel = context.getTemporaryDir().toPath().resolve("quarkus-app-model.dat");
         try (ObjectOutputStream out = new ObjectOutputStream(Files.newOutputStream(serializedModel))) {
             out.writeObject(appModel);
         }
         return serializedModel;
+    }
+
+    public static String getClassesDir(SourceSet sourceSet, AbstractTask context) {
+        final Set<String> sourcePaths = new HashSet<>();
+        for (File sourceDir : sourceSet.getAllJava().getSrcDirs()) {
+            sourcePaths.add(sourceDir.getAbsolutePath());
+        }
+
+        FileCollection classesDirs = sourceSet.getOutput().getClassesDirs();
+        Set<File> classDirFiles = classesDirs.getFiles();
+        if (classDirFiles.size() == 1) {
+            return classesDirs.getAsPath();
+        }
+
+        Path classesDir = null;
+        final Iterator<File> i = classDirFiles.iterator();
+        int dirCount = 0;
+        while (i.hasNext()) {
+            final File next = i.next();
+            if (!next.exists()) {
+                continue;
+            }
+            try {
+                switch (dirCount++) {
+                    case 0:
+                        classesDir = next.toPath();
+                        break;
+                    case 1:
+                        //there does not seem to be any sane way of dealing with multiple output dirs, as there does not seem
+                        //to be a way to map them. We will need to address this at some point, but for now we just stick them
+                        //all in a temp dir
+                        final Path tmpClassesDir = context.getTemporaryDir().toPath().resolve("quarkus-app-classes");
+                        if (Files.exists(tmpClassesDir)) {
+                            IoUtils.recursiveDelete(tmpClassesDir);
+                        }
+                        IoUtils.copy(classesDir, tmpClassesDir);
+                        classesDir = tmpClassesDir;
+                    default:
+                        IoUtils.copy(next.toPath(), classesDir);
+
+                }
+            } catch (IOException e) {
+                throw new UncheckedIOException(ERROR_COLLECTING_PROJECT_CLASSES, e);
+            }
+        }
+        if (classesDir == null) {
+            throw new IllegalStateException("Failed to locate classes directory in the project");
+        }
+        return classesDir.toString();
     }
 }

--- a/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusTestConfig.java
+++ b/devtools/gradle/src/main/java/io/quarkus/gradle/tasks/QuarkusTestConfig.java
@@ -27,7 +27,7 @@ public class QuarkusTestConfig extends QuarkusTask {
                     .toAbsolutePath()
                     .toString();
 
-            final Path serializedModel = QuarkusGradleUtils.serializeAppModel(deploymentDeps);
+            final Path serializedModel = QuarkusGradleUtils.serializeAppModel(deploymentDeps, this);
 
             for (Test test : getProject().getTasks().withType(Test.class)) {
                 final Map<String, Object> props = test.getSystemProperties();

--- a/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/DirectoryClassPathElement.java
+++ b/independent-projects/bootstrap/core/src/main/java/io/quarkus/bootstrap/classloading/DirectoryClassPathElement.java
@@ -22,6 +22,7 @@ public class DirectoryClassPathElement extends AbstractClassPathElement {
     private final Path root;
 
     public DirectoryClassPathElement(Path root) {
+        assert root != null : "root is null";
         this.root = root;
     }
 
@@ -101,5 +102,10 @@ public class DirectoryClassPathElement extends AbstractClassPathElement {
     @Override
     public void close() throws IOException {
         //noop
+    }
+
+    @Override
+    public String toString() {
+        return root.toAbsolutePath().toString();
     }
 }

